File: ios-plugin/calendar/CalendarContextPlugin.swift
import Foundation
import Capacitor
import EventKit

@objc(CalendarContextPlugin)
public class CalendarContextPlugin: CAPPlugin {
    private let eventStore = EKEventStore()
    private let workStartMinute = 8 * 60
    private let workEndMinute = 20 * 60

    @objc func getPermissionStatus(_ call: CAPPluginCall) {
        call.resolve(["state": permissionState()])
    }

    @objc func requestPermission(_ call: CAPPluginCall) {
        let complete: (Bool, Error?) -> Void = { granted, error in
            if let error = error {
                call.reject("Calendar permission failed: \(error.localizedDescription)")
                return
            }
            call.resolve(["state": granted ? "granted" : "denied"])
        }

        if #available(iOS 17.0, *) {
            eventStore.requestFullAccessToEvents(completion: complete)
        } else {
            eventStore.requestAccess(to: .event, completion: complete)
        }
    }

    @objc func getDailyAggregates(_ call: CAPPluginCall) {
        guard permissionState() == "granted" else {
            call.resolve(["state": permissionState(), "days": []])
            return
        }
        guard let startText = call.getString("startDate"),
              let endText = call.getString("endDate"),
              let start = dateFromDay(startText),
              let endDay = dateFromDay(endText),
              let exclusiveEnd = Calendar.current.date(byAdding: .day, value: 1, to: endDay),
              start <= endDay else {
            call.reject("startDate and endDate must use yyyy-MM-dd")
            return
        }

        let predicate = eventStore.predicateForEvents(withStart: start, end: exclusiveEnd, calendars: nil)
        let events = eventStore.events(matching: predicate).filter { !$0.isAllDay }
        var days: [[String: Any]] = []
        var cursor = start

        while cursor < exclusiveEnd {
            guard let next = Calendar.current.date(byAdding: .day, value: 1, to: cursor) else { break }
            let intervals = events.compactMap { event -> (Date, Date)? in
                let clippedStart = max(event.startDate, cursor)
                let clippedEnd = min(event.endDate, next)
                return clippedEnd > clippedStart ? (clippedStart, clippedEnd) : nil
            }
            days.append(aggregateDay(dayStart: cursor, intervals: intervals))
            cursor = next
        }

        call.resolve(["state": "granted", "days": days])
    }

    private func permissionState() -> String {
        let status = EKEventStore.authorizationStatus(for: .event)
        if #available(iOS 17.0, *) {
            switch status {
            case .fullAccess, .authorized: return "granted"
            case .writeOnly, .restricted, .denied: return "denied"
            case .notDetermined: return "not_determined"
            @unknown default: return "unavailable"
            }
        }
        switch status {
        case .authorized: return "granted"
        case .restricted, .denied: return "denied"
        case .notDetermined: return "not_determined"
        default: return "unavailable"
        }
    }

    private func dateFromDay(_ value: String) -> Date? {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone.current
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.date(from: value)
    }

    private func minuteOfDay(_ date: Date, dayStart: Date) -> Int {
        max(0, min(1440, Int(date.timeIntervalSince(dayStart) / 60)))
    }

    private func aggregateDay(dayStart: Date, intervals: [(Date, Date)]) -> [String: Any] {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone.current
        formatter.dateFormat = "yyyy-MM-dd"

        let minuteIntervals = intervals.map {
            (minuteOfDay($0.0, dayStart: dayStart), minuteOfDay($0.1, dayStart: dayStart))
        }.filter { $0.1 > $0.0 }.sorted { $0.0 < $1.0 }

        var merged: [(Int, Int)] = []
        for interval in minuteIntervals {
            if let last = merged.last, interval.0 <= last.1 {
                merged[merged.count - 1] = (last.0, max(last.1, interval.1))
            } else {
                merged.append(interval)
            }
        }

        let busyMinutes = merged.reduce(0) { $0 + ($1.1 - $1.0) }
        let longestMeeting = minuteIntervals.map { $0.1 - $0.0 }.max() ?? 0
        let currentMinute = Calendar.current.isDate(dayStart, inSameDayAs: Date())
            ? minuteOfDay(Date(), dayStart: dayStart)
            : workStartMinute
        let effectiveWorkStart = min(workEndMinute, max(workStartMinute, currentMinute))
        let workBusy = merged.compactMap { interval -> (Int, Int)? in
            let start = max(effectiveWorkStart, interval.0)
            let end = min(workEndMinute, interval.1)
            return end > start ? (start, end) : nil
        }
        var openCursor = effectiveWorkStart
        var longestOpenStart: Int? = effectiveWorkStart < workEndMinute ? effectiveWorkStart : nil
        var longestOpenMinutes = 0
        for interval in workBusy {
            if interval.0 > openCursor && interval.0 - openCursor > longestOpenMinutes {
                longestOpenStart = openCursor
                longestOpenMinutes = interval.0 - openCursor
            }
            openCursor = max(openCursor, interval.1)
        }
        if workEndMinute - openCursor > longestOpenMinutes {
            longestOpenStart = openCursor
            longestOpenMinutes = workEndMinute - openCursor
        }

        return [
            "date": formatter.string(from: dayStart),
            "busyMinutes": busyMinutes,
            "meetingCount": minuteIntervals.count,
            "longestMeetingMinutes": longestMeeting,
            "firstEventMinute": minuteIntervals.first?.0 ?? NSNull(),
            "lastEventMinute": minuteIntervals.last?.1 ?? NSNull(),
            "longestOpenStartMinute": longestOpenStart ?? NSNull(),
            "longestOpenMinutes": longestOpenMinutes,
        ]
    }
}

File: android-plugin/calendar/CalendarContextPlugin.kt
package com.looma.plugins

import android.Manifest
import android.content.pm.PackageManager
import android.provider.CalendarContract
import androidx.core.content.ContextCompat
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.PermissionState
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.annotation.PermissionCallback
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId

@CapacitorPlugin(
    name = "CalendarContext",
    permissions = [Permission(strings = [Manifest.permission.READ_CALENDAR], alias = "calendar")]
)
class CalendarContextPlugin : Plugin() {
    private val workStartMinute = 8 * 60
    private val workEndMinute = 20 * 60

    @PluginMethod
    fun getPermissionStatus(call: PluginCall) {
        call.resolve(JSObject().put("state", permissionState()))
    }

    @PluginMethod
    fun requestPermission(call: PluginCall) {
        if (getPermissionState("calendar") == PermissionState.GRANTED) {
            call.resolve(JSObject().put("state", "granted"))
            return
        }
        requestPermissionForAlias("calendar", call, "calendarPermissionCallback")
    }

    @PermissionCallback
    private fun calendarPermissionCallback(call: PluginCall) {
        call.resolve(JSObject().put("state", permissionState()))
    }

    @PluginMethod
    fun getDailyAggregates(call: PluginCall) {
        if (permissionState() != "granted") {
            call.resolve(JSObject().put("state", permissionState()).put("days", JSArray()))
            return
        }
        val startDate = call.getString("startDate")
            ?.let { runCatching { LocalDate.parse(it) }.getOrNull() }
            ?: run {
                call.reject("startDate and endDate must use yyyy-MM-dd")
                return
            }
        val endDate = call.getString("endDate")
            ?.let { runCatching { LocalDate.parse(it) }.getOrNull() }
            ?: run {
                call.reject("startDate and endDate must use yyyy-MM-dd")
                return
            }
        if (startDate.isAfter(endDate)) {
            call.reject("startDate and endDate must use yyyy-MM-dd")
            return
        }

        val zone = ZoneId.systemDefault()
        val rangeStart = startDate.atStartOfDay(zone).toInstant().toEpochMilli()
        val rangeEnd = endDate.plusDays(1).atStartOfDay(zone).toInstant().toEpochMilli()
        val uri = CalendarContract.Instances.CONTENT_URI.buildUpon()
            .appendPath(rangeStart.toString())
            .appendPath(rangeEnd.toString())
            .build()
        val projection = arrayOf(
            CalendarContract.Instances.BEGIN,
            CalendarContract.Instances.END,
            CalendarContract.Instances.ALL_DAY,
        )
        val byDay = mutableMapOf<LocalDate, MutableList<Pair<Long, Long>>>()

        context.contentResolver.query(uri, projection, null, null, null)?.use { cursor ->
            val beginIndex = cursor.getColumnIndexOrThrow(CalendarContract.Instances.BEGIN)
            val endIndex = cursor.getColumnIndexOrThrow(CalendarContract.Instances.END)
            val allDayIndex = cursor.getColumnIndexOrThrow(CalendarContract.Instances.ALL_DAY)
            while (cursor.moveToNext()) {
                if (cursor.getInt(allDayIndex) == 1) continue
                val begin = cursor.getLong(beginIndex)
                val end = cursor.getLong(endIndex)
                if (end <= begin) continue
                var day = Instant.ofEpochMilli(begin).atZone(zone).toLocalDate()
                val finalDay = Instant.ofEpochMilli(end - 1).atZone(zone).toLocalDate()
                while (!day.isAfter(finalDay)) {
                    byDay.getOrPut(day) { mutableListOf() }.add(Pair(begin, end))
                    day = day.plusDays(1)
                }
            }
        }

        val days = JSArray()
        var day = startDate
        while (!day.isAfter(endDate)) {
            days.put(aggregateDay(day, byDay[day] ?: emptyList(), zone))
            day = day.plusDays(1)
        }
        call.resolve(JSObject().put("state", "granted").put("days", days))
    }

    private fun permissionState(): String = when {
        ContextCompat.checkSelfPermission(context, Manifest.permission.READ_CALENDAR) ==
            PackageManager.PERMISSION_GRANTED -> "granted"
        getPermissionState("calendar") == PermissionState.PROMPT -> "not_determined"
        else -> "denied"
    }

    private fun aggregateDay(
        date: LocalDate,
        rawIntervals: List<Pair<Long, Long>>,
        zone: ZoneId,
    ): JSObject {
        val dayStart = date.atStartOfDay(zone).toInstant().toEpochMilli()
        val dayEnd = date.plusDays(1).atStartOfDay(zone).toInstant().toEpochMilli()
        val intervals = rawIntervals.mapNotNull { raw ->
            val start = maxOf(dayStart, raw.first)
            val end = minOf(dayEnd, raw.second)
            if (end <= start) null else Pair(
                ((start - dayStart) / 60_000).toInt(),
                ((end - dayStart) / 60_000).toInt(),
            )
        }.sortedBy { it.first }

        val merged = mutableListOf<Pair<Int, Int>>()
        intervals.forEach { interval ->
            val last = merged.lastOrNull()
            if (last != null && interval.first <= last.second) {
                merged[merged.lastIndex] = Pair(last.first, maxOf(last.second, interval.second))
            } else {
                merged.add(interval)
            }
        }
        val busyMinutes = merged.sumOf { it.second - it.first }
        val longestMeeting = intervals.maxOfOrNull { it.second - it.first } ?: 0
        val currentMinute = if (date == LocalDate.now(zone)) {
            val now = java.time.ZonedDateTime.now(zone)
            now.hour * 60 + now.minute
        } else {
            workStartMinute
        }
        val effectiveWorkStart = minOf(workEndMinute, maxOf(workStartMinute, currentMinute))
        val workBusy = merged.mapNotNull {
            val start = maxOf(effectiveWorkStart, it.first)
            val end = minOf(workEndMinute, it.second)
            if (end > start) Pair(start, end) else null
        }
        var cursor = effectiveWorkStart
        var longestOpenStart: Int? = if (effectiveWorkStart < workEndMinute) effectiveWorkStart else null
        var longestOpenMinutes = 0
        workBusy.forEach { interval ->
            if (interval.first - cursor > longestOpenMinutes) {
                longestOpenStart = cursor
                longestOpenMinutes = interval.first - cursor
            }
            cursor = maxOf(cursor, interval.second)
        }
        if (workEndMinute - cursor > longestOpenMinutes) {
            longestOpenStart = cursor
            longestOpenMinutes = workEndMinute - cursor
        }

        return JSObject()
            .put("date", date.toString())
            .put("busyMinutes", busyMinutes)
            .put("meetingCount", intervals.size)
            .put("longestMeetingMinutes", longestMeeting)
            .put("firstEventMinute", intervals.firstOrNull()?.first)
            .put("lastEventMinute", intervals.lastOrNull()?.second)
            .put("longestOpenStartMinute", longestOpenStart)
            .put("longestOpenMinutes", longestOpenMinutes)
    }
}
